pandas/io/sql.py

```diff
@@ -2024,6 +2065,296 @@ def _create_sql_schema(
+        if not isinstance(sql, str):
+            raise TypeError("Query must be a string unless using sqlalchemy.")
+        args = [] if params is None else [params]
+        cur = self.con.cursor()
+        try:
+            cur.execute(sql, *args)
+            return cur
+        except Exception as exc:
+            try:
+                self.con.rollback()
+            except Exception as inner_exc:  # pragma: no cover
+                ex = DatabaseError(
+                    f"Execution failed on sql: {sql}\n{exc}\nunable to rollback"
+                )
+                raise ex from inner_exc
+
+            ex = DatabaseError(f"Execution failed on sql '{sql}': {exc}")
+            raise ex from exc
+
+    def read_table(
+        self,
+        table_name: str,
+        index_col: str | list[str] | None = None,
+        coerce_float: bool = True,
+        parse_dates=None,
+        columns=None,
+        schema: str | None = None,
+        chunksize: int | None = None,
+        dtype_backend: DtypeBackend | Literal["numpy"] = "numpy",
+    ) -> DataFrame | Iterator[DataFrame]:
+        """
+        Read SQL database table into a DataFrame.
+
+        Parameters
+        ----------
+        table_name : str
+            Name of SQL table in database.
+        schema : string, default None
+            Name of SQL schema in database to read from
+
+        Returns
+        -------
+        DataFrame
+
+        See Also
+        --------
+        pandas.read_sql_table
+        SQLDatabase.read_query
+
+        """
+        if coerce_float is not True:
+            raise NotImplementedError(
+                "'coerce_float' is not implemented for ADBC drivers"
+            )
+        if chunksize:
+            raise NotImplementedError("'chunksize' is not implemented for ADBC drivers")
+
+        if columns:
+            if index_col:
+                index_select = maybe_make_list(index_col)
+            else:
+                index_select = []
+            to_select = index_select + columns
+            select_list = ", ".join(f'"{x}"' for x in to_select)
+        else:
+            select_list = "*"
+        if schema:
+            stmt = f"SELECT {select_list} FROM {schema}.{table_name}"
+        else:
+            stmt = f"SELECT {select_list} FROM {table_name}"
+
+        mapping: type[ArrowDtype] | None | Callable
+        if dtype_backend == "pyarrow":
+            mapping = ArrowDtype
+        elif dtype_backend == "numpy_nullable":
+            from pandas.io._util import _arrow_dtype_mapping
+
+            mapping = _arrow_dtype_mapping().get
+        elif using_pyarrow_string_dtype():
+            from pandas.io._util import arrow_string_types_mapper
+
+            arrow_string_types_mapper()
+        else:
+            mapping = None
+
+        with self.con.cursor() as cur:
+            cur.execute(stmt)
+            df = cur.fetch_arrow_table().to_pandas(types_mapper=mapping)
```

Comment:
So a remaining issue seems to be that the postgres driver is pulling back `int32` values for INTEGERs and `datetime64[us]` dtypes for datetimes, which is natural given postgres data types. I'm not sure if we still want to map those back to int64 / nanosecond-based datetime to match the pandas datatypes, and if so how we should be bolting that onto the existing types_mapper argument, if at all.

@jorisvandenbossche  @phofl if you have any guidance would greatly appreciate it

Reply:
I would personally lean with keeping those data types as they are coming from the postgres ADBC driver. For the integers, if they are already int32 in the arrow table, casting to int64 doesn't give much benefit. And for datetimes, we should nowadays supports the other resolutions well, so I would also keep the resolution (we do the same now in Parquet reading in pyarrow if the parquet timestamp are non-nanosecond)

Reply:
The nuance I failed to mention before was that this only applies to the postgres ADBC driver. Looks like the sqlite ADBC driver happily returns 64 bit integers from sqlite, assumedly since sqlite doesn't have sized integers. 

So potentially confusing that you may get different types back depending on the driver, but there might not be a better solution than understanding what types the database you are using offers in the first place

Reply:
And looks like the temporal unit is dependent upon the version of pyarrow...

Reply:
But postgres does have a BIGINT (https://www.postgresql.org/docs/current/datatype-numeric.html), so my expectation would be that an int64 column roundtrips correctly? Or is that not the case? 
(it might just not be the typical int size used by many databases)

Reply:
For roundtripping where you created the data in pandas and wrote to the database there is no problem. The issue is with tables created in postgres

```sh
CREATE TABLE test_table (column1 INTEGER);
INSERT INTO test_table VALUES ((1));
```

`df = pd.read_table("SELECT * FROM test_table", con=<adbc_conn>` yields int32 whereas using a SQLAlchemy conn yields int64

I _think_ most non-pandas tools would naturally create `INTEGER` types unless 64 bit is really required, so just a difference in behavior when interacting outside of the pandas roundtrip system

Reply:
If the database uses INTEGER, I think it's fine that we read that as int32. In the end, that is more correct, and the only reason that we always have used int64 is because that is the default integer type in pandas, and we just get python tuples with python ints from the driver, and thus infer our default int64 when converting this list of tuples to a DataFrame (we don't actually check the original type in the database for numeric data, `_harmonize_columns` only checks for datetime-like columns)